src/core/ParticleIterator.hpp

```diff
@@ -16,6 +16,10 @@ struct ParticleIterator : public boost::iterator_facade<
     }
   }
 
+  operator Particle *() const {
```

Comment:
ParticleIterator has already an operator* which returns Particle&. You cannot overload it by return type. ESPResSo, however, uses functions which take pointers to particles everywhere (e.g. above: do_nonbonded, add_single_particle_force, ...).